### DISCUSSION TITLE
wasm_loader.js file not found - Cancelling startup

### DISCUSSION THREAD

**Discussion by PythonHacker24** [kanidm/kanidm#2286] [General]
I am facing an issue while setting up Kanidm as per the developer guide. This is happening with the builtin web-ui when I run ./run_insecure_dev_server.sh. It throws an error of: 
`[error]: Failed to generate integrity hash for wasmloader_admin.js - cancelling startup! | event_tag_id: 1 | err: "Can't find \"../web_ui/pkg/wasmloader_admin.js\" to generate file hash"`
Can anyone please help me with this?

**Comment by Firstyear** [comment]
Are you running the script from the correct directory? You need to cd to server/daemon first.

**Reply by PythonHacker24** [reply]
Yes I am doing it in the correct directory as per the developer guide

**Reply by yaleman** [reply]
Check that `wasmloader_admin.js` is at `./server/web_ui/pkg/wasmloader_admin.js` from the base of the repo.

Try running `make webui` from the root of the repository either way, it'll build the WASM files.

**Reply by PythonHacker24** [reply]
The `wasmloader_admin.js` is in the repo, but when `build_wasm_dev.sh` is executed, the `wasmloader_admin.js` gets deleted and after I run `./run_insecure_dev_server.sh` I get a error that it doesn't exist

**Reply by yaleman** [reply]
When you run `build_wasm_dev.sh` what do you see between `building admin` and `done building admin` - I'm guessing something is failing at that point.

**Reply by PythonHacker24** [reply]
`[INFO]: ⬇️  Installing wasm-bindgen...
Error: Not able to find or install a local wasm-bindgen.
Caused by: Not able to find or install a local wasm-bindgen.` Is this it, looks suspicious

**Reply by PythonHacker24** [reply]
This came after building admin

**Reply by yaleman** [reply]
That'll do it, try `cargo install wasm-pack` and if that's missing then .. try [cargo install -f wasm-bindgen-cli](https://github.com/rustwasm/wasm-pack/issues/1138)?

**Reply by PythonHacker24** [reply]
I have tried the `cargo install -f wasm-bindgen-cli`. I have the `wasm-bindgen-cli` installed in my system. I can even run `wasm-bindgen` on my terminal with output:
`Invalid arguments.
Usage:
    wasm-bindgen [options] <input>
    wasm-bindgen -h | --help
    wasm-bindgen -V | --version`
Which looks like it had been installed. The `./build_wasm_dev.sh` still points that the `local wasm-bindgen` is not installed. 

While the building admin part starts: `indivisual_build.sh` is executed where `wasm-pack build ${BUILD_FLAGS} --no-typescript --target web --mode no-install --no-pack` is throwing an error when the `${BUILD_FLAG}` is substituted with `--release`

**Reply by PythonHacker24** [reply]
When I run `wasm-pack build --release --no-typescript --target web --mode no-install --no-pack` on the terminal in the directory of `kanidm/server/web_ui` 

`Error: failed to parse manifest: /Users/adityapatil/Open-Source/kanidm/Cargo.toml
Caused by: failed to parse manifest: /Users/adityapatil/Open-Source/kanidm/Cargo.toml
Caused by: TOML parse error at line 1, column 1`
The errors gets town to:
`[profile.release] -> missing field "package"`

And My Cargo.toml contains: 
`[profile.release]`
`debug = true`
`lto = "thin"`

I can see that there is no package field here.

**Reply by yaleman** [reply]
You shouldn't be running it in `server/web_ui` that's not a package - `admin`, `login_flows`, `shared` and `user` are the packages. The Cargo.toml you're referencing is the workspace one, which is irrelevant in context.

The error's coming from within the packages themselves, we're not calling it directly from our scripts - what OS/distribution are you building it on? I know it'll fail in an aarch64 linux container because they don't support that combination.

**Reply by PythonHacker24** [reply]
I am using MacOS Sonoma on M2 processor. Is there any issue with ARM64 chips?

**Reply by yaleman** [reply]
That's what I use M2 max, I'm on Ventura but that *shouldn't* matter. Try ensuring that you've got the latest tools?

I'm running wasm-pack 0.12.1 and rust 1.73.

**Reply by PythonHacker24** [reply]
My wasm-pack version `wasm-pack 0.12.1` 
My rust version `1.73.0`

The guide I am using: `https://kanidm.github.io/kanidm/stable/DEVELOPER_README.html`
I have followed step by step to the process given there. Let me add what I did:

1. `cargo install wasm-pack` with Xcode existing + `rustup`
2. `KANIDM_BUILD_PROFILE=release_suse_generic cargo build --release --bin kanidmd`
3. `cd server/web_ui/`
`./build_wasm_dev.sh`
4. `cd server/daemon`
`./run_insecure_dev_server.sh`
And I crash ....

**Reply by yaleman** [reply]
I really don't know what you're hitting then, this seems to be a build tooling issue outside the source code of Kanidm - if you're still getting the following error (since you haven't reported what's breaking, currently, I can only assume it's that):

> [INFO]: ⬇️  Installing wasm-bindgen... Error: Not able to find or install a local wasm-bindgen. Caused by: Not able to find or install a local wasm-bindgen.

Then it's a `wasm-bindgen` or other tooling issue that I can't identify remotely. Focus on step #3 because ... there's no point continuing if that's not succeeding.

**Reply by PythonHacker24** [reply]
Okay will try to isolate the point of the issue and will fix it.

**Reply by PythonHacker24** [reply]
![image](https://github.com/kanidm/kanidm/assets/72488360/89c5a622-1e54-4148-b7b9-6aac77269e09)

**Reply by PythonHacker24** [reply]
The command `wasm-pack build --release --no-typescript --target web --mode no-install --no-pack` was ran from the `server/web_ui` directory

**Reply by yaleman** [reply]
Why run wasm-pack from there though? I've already said that's wrong - and the build script doesn't do it that way either  - you need to run it from one of the package directories. You have to run it from admin or user or shared or login_flows:

<img width="1067" alt="Screenshot 2023-11-04 at 00 35 03" src="https://github.com/kanidm/kanidm/assets/168188/80d38348-32e8-430f-abb8-e60975261c22">

notice the `cd <folder>`:

```shell
cd admin <=- <=- 
echo "building admin"
../individual_build.sh || exit 1
cd ..
echo "done building admin"
```

**Reply by PythonHacker24** [reply]
Yes I did execute via the script, I did it earlier actually, I was just testing around it

**Reply by PythonHacker24** [reply]
![image](https://github.com/kanidm/kanidm/assets/72488360/7a8a2989-64e0-4214-8286-1969e15e4700)

**Reply by PythonHacker24** [reply]
Then I tried going to the admin file and run the `wasm-pack build --release --no-typescript --target web --mode no-install --no-pack` 
Now I am upto:
![image](https://github.com/kanidm/kanidm/assets/72488360/3dff1e88-6dab-4da3-9693-199df13c545b)

**Reply by PythonHacker24** [reply]
And all gets down to not able to find the local wasm-bindgen
![image](https://github.com/kanidm/kanidm/assets/72488360/28e33430-691a-443e-8fe8-fe1d83c71415)
The local does exists

**Reply by yaleman** [reply]
> Yes I did execute via the script, I did it earlier actually, I was just testing around it

Ok, because running the wrong thing and expecting it to work just confuses the situation 😄 

Did you try running the command from the [linked thread](https://github.com/rustwasm/wasm-pack/issues/1138#issuecomment-1439089704) and checking all the output? It seems like it doesn't *always* use the local version:

```shell
RUST_LOG=DEBUG wasm-pack build --release --no-typescript --target web --mode no-install --no-pack 2>&1 | grep -i installed
[INFO  wasm_pack::install] Checking installed `wasm-bindgen` version == expected version: 0.2.87 == 0.2.87
```

**Reply by PythonHacker24** [reply]
![image](https://github.com/kanidm/kanidm/assets/72488360/742776d0-be5e-4d16-a86d-3ae04265221f)
Here we go with the output. Are the versions causing some issues?

**Reply by PythonHacker24** [reply]
Hey, I got it done and installed!!!! 

The wasm-bindgen-cli version was the whole issue here. I installed version `0.2.87` as it moved forward the script!

Should I work on the `build_wasm_dev.sh` and include the instructions to cross-check the desired version of the `wasm-bindgen-cli` and install the correct one and proceed with the wasm build as well as update the document for the same?

And a big thank you for helping me out!

**Reply by Firstyear** [reply]
If the version was the problem, should we actually have a hard-gated check on the version to ensure it's up to date? Rather than updating docs, we should strictly enforce our requirements.

**Reply by yaleman** [reply]
Looks like we need to provide troubleshooting documentation, *we* don't set that required version, wasm-pack does, and it looks like there was a mismatch between the versions of the various tools @PythonHacker24 had on their machine.

I just updated to v0.2.88 and the WASM built/worked fine, so it's not a hard limit on 0.2.87... there's a *lot* of threads around about this exact error and needing to force-reinstall `wasm-bindgen-cli` being the fix.

```shell
15:54:50 ➜ wasm-bindgen --version
wasm-bindgen 0.2.88

15:56:21 ➜ wasm-pack --version
wasm-pack 0.12.1
```

**Reply by PythonHacker24** [reply]
I got the same version for the wasm-pack: `wasm-pack 0.12.1`

**Reply by PythonHacker24** [reply]
And now with wasm-bindgen-cli version: `0.2.87` the web-ui is up and running. 

Is this gonna be updated for the troubleshooting docs: `https://kanidm.github.io/kanidm/stable/troubleshooting.html` 
for the specific error in the case of 
`[error]: Failed to generate integrity hash for wasmloader_admin.js - cancelling startup! | event_tag_id: 1| err: "Can't find \"../web_ui/pkg/wasmloader_admin.js\" to generate file hash"` 
issue and well as finding out the 
`[INFO]: ⬇️  Installing wasm-bindgen...` 
`Error: Not able to find or install a local wasm-bindgen.` 
`Caused by: Not able to find or install a local wasm-bindgen.`

**Reply by yaleman** [reply]
It's not an issue relating to Kanidm, our configuration or otherwise, it's a `wasm-pack` / `wasm-bindgen-cli` issue, which reinstalling the tool fixed. 

We don't maintain those packages. Try upgrading again and it'll probably work. If not, then I'd suggest a bug report in their repository which they might be able to help with, and we can then reference in our docs.

**Reply by yaleman** [reply]
> [error]: Failed to generate integrity hash for wasmloader_admin.js - cancelling startup! | event_tag_id: 1| err: "Can't find \"../web_ui/pkg/wasmloader_admin.js\" to generate file hash"

This error must have occurred after you attempted to build the webui and it failed - the repository always has these files compiled and available.

**Reply by PythonHacker24** [reply]
Yes, they are there in the repository, I am not sure my the versioning affected the build in the case. It must be the wasm-pack or wasm-buildgen-cli which might be causing an error. It was not the kanidm scope, the outside dependency. I would go for an issue on wasm-pack and wasm-buildgen-cli after investigating this further. We don't want anyone else to go through this error and work out the versioning error again.
Can we go for providing a temporary solution for this issue while the external wasm dependency issue is fixed? Just like a suggestion to try out this error is for anyone who sets this up before the issue is resolved.

**Reply by PythonHacker24** [reply]
I am aware that the versioning is not only the issue here as it worked on your system but I think it might be the case of the stability of the `0.2.88` which may have caused issues for new installations. I would test this in other macs as well and find out if that is an os specific bug, any dependency collisions, or other things. Till now, it looks like the `0.2.87` is suitable for new installations.

**Reply by yaleman** [reply]
> I am not sure my the versioning affected the build in the case. It must be the wasm-pack or wasm-buildgen-cli which might be causing an error. 

The build script for the WASM packages clears the package directory, so as to avoid having stale artifacts there. If you run it, and the build fails, then the files won't be in the package dir, so you have to sort **that** before trying to run the server... 

The build failed due to **_something_** in a `wasm-pack` command failing to complete, that was resolved by you reinstalling the tooling. I'd suggest PR'ing a change that links to [this thread in the wasm-pack repository](https://github.com/rustwasm/wasm-pack/issues/1138) saying to force-reinstall `wasm-bindgen-cli` and `wasm-pack` if they get the error you saw.

**Reply by PythonHacker24** [reply]
So I should add a suggestion function in the build bash script, about when a similar failure like the one I had occurs, they must try to force reinstall the `wasm-bindgen-cli` and `wasm-pack` as well as providing a [link to this thread](https://github.com/rustwasm/wasm-pack/issues/1138). Am I right?

**Reply by yaleman** [reply]
No need to over complicate it, I've done [the PR](https://github.com/kanidm/kanidm/pull/2291).

**Reply by PythonHacker24** [reply]
I wanted to do that ....

**Reply by yaleman** [reply]
But now we all get to go back to working on the fun Rust things we were working on ~24 hours ago 😄

**Reply by PythonHacker24** [reply]
Cool, I would like to dive deep into this project and make further contributions ✨
